src/python/grpcio/tests/unit/framework/common/test_control.py

```diff
@@ -72,19 +73,31 @@ def control(self):
       if self._fail:
         raise Defect()
 
-      while self._paused:
+      while self._pause:
+        self._paused = True
+        self._condition.notify_all()
         self._condition.wait()
+      self._paused = False
 
   @contextlib.contextmanager
   def pause(self):
     """Pauses code under control while controlling code is in context."""
     with self._condition:
-      self._paused = True
+      self._pause = True
     yield
     with self._condition:
```

Comment:
That would be really bad because the way this object is used is
```
with control.pause():
  <set in motion the behavior under test that would lead to the system under test being paused>
  <either call control.block_until_paused or don't>
  <maybe do some other stuff>
```
. With the semantics you suggest, that code would have to be changed to
```
<set in motion the behavior under test that would lead to the system under test being paused>
with control.pause():
  <either call control.block_until_paused or don't>
  <maybe do some other stuff>
```
and that would create a race condition in which the system under test could pass the control point (where it calls `control.control()`) _before_ the test system enters the `control.pause()` context.

Concretely, look at the way this is used in expiration tests: we say that the service-side application should pause, then we invoke the RPC, then we chill out until we see the RPC time out, and then we leave the pause context and allow the service-side application to return from its RPC method handler. With the behavior you propose, the test would have to invoke the RPC before entering the pause context and the service-side application could finish servicing the RPC method before the test enters the pause context.

Right?